# Tarea 5

## Motor Trend Car Road Tests

Julia Hernández Cárdenas

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score
import statsmodels.api as sm
from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [2]:
motor = pd.read_excel("Motor Trend Car Road Tests.xlsx")
motor.dropna()
motor.drop_duplicates()
motor = motor.drop(columns = ["model"])

**1.1** Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.

1. Calcula el R2 e interpreta los signos de los betas.
2. Realiza un train-test-split donde se use el 40% de los datos para entrenar. 
3. Calcula el R2 de entrenamiento y de prueba.
4. Añade regularización L2 con un hiperparámetro lambda decidido por ti. 
5. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.

In [3]:
x = motor.drop(columns=["mpg"])
y = motor["mpg"]

**1.**

In [4]:
x = sm.add_constant(x)
ols = sm.OLS(y, x)
results = ols.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.869
Model:                            OLS   Adj. R-squared:                  0.807
Method:                 Least Squares   F-statistic:                     13.93
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           3.79e-07
Time:                        16:48:14   Log-Likelihood:                -69.855
No. Observations:                  32   AIC:                             161.7
Df Residuals:                      21   BIC:                             177.8
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.3034     18.718      0.657      0.518     -26.623      51.229
cyl           -0.1114      1.045     -0.107      0.916      -2.285       2.062
disp           0.0133      0.018      0.747      0.463      -0.024       0.050
hp            -0.0215      0.022     -0.987      0.335      -0.067       0.024
drat           0.7871      1.635      0.481      0.635      -2.614       4.188
wt            -3.7153      1.894     -1.961      0.063      -7.655       0.224
qsec           0.8210      0.731      1.123      0.274      -0.699       2.341
vs             0.3178      2.105      0.151      0.881      -4.059       4.694
am             2.5202      2.057      1.225      0.234      -1.757       6.797
gear           0.6554      1.493      0.439      0.665      -2.450       3.761
carb          -0.1994      0.829     -0.241      0.812      -1.923       1.524
==============================================================================
Omnibus:                        1.907   Durbin-Watson:                   1.861
Prob(Omnibus):                  0.385   Jarque-Bera (JB):                1.747
Skew:                           0.521   Prob(JB):                        0.418
Kurtosis:                       2.526   Cond. No.                     1.22e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.22e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

El R^2 es de 0.869, y las betas salieron tanto positivas como negativas, indicando que hay factores (como am, qsec y gear que si influyen, y valores (como wt y carb) que no influyen tan bien

**2.**

In [5]:
Xtrain, Xtest, Ytrain, Ytest = train_test_split(x, y, train_size = 0.4, random_state = 137)

**3.**

In [6]:
scaler = StandardScaler().fit(Xtrain)
X_scaled = scaler.transform(Xtrain)

Entrenamiento

In [7]:
X_modelo2 = sm.add_constant(X_scaled)
ols2 = sm.OLS(Ytrain, X_modelo2)
results2 = ols2.fit()
results2.summary()

/Users/juliahernandez/anaconda3/lib/python3.11/site-packages/scipy/stats/_stats_py.py:1736: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.990
Model:                            OLS   Adj. R-squared:                  0.891
Method:                 Least Squares   F-statistic:                     9.981
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.242
Time:                        16:48:14   Log-Likelihood:                -9.9811
No. Observations:                  12   AIC:                             41.96
Df Residuals:                       1   BIC:                             47.30
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.4333      0.556     31.361      0.020      10.370      24.497
x1          1.005e-13   1.26e-13      0.795      0.572   -1.51e-12    1.71e-12
x2            40.7571     54.010      0.755      0.588    -645.499     727.013
x3            -2.4635      3.021     -0.815      0.565     -40.854      35.927
x4           -10.1581     19.484     -0.521      0.694    -257.727     237.411
x5            14.1135     13.469      1.048      0.485    -157.022     185.249
x6            -0.8005      3.106     -0.258      0.839     -40.266      38.665
x7             1.5727      3.475      0.453      0.729     -42.583      45.729
x8            32.4926     41.262      0.787      0.575    -491.795     556.780
x9            36.1394     42.669      0.847      0.553    -506.022     578.301
x10          -38.3551     43.625     -0.879      0.541    -592.668     515.958
x11           11.9707     17.482      0.685      0.618    -210.157     234.098
==============================================================================
Omnibus:                        0.259   Durbin-Watson:                   2.354
Prob(Omnibus):                  0.879   Jarque-Bera (JB):                0.413
Skew:                           0.212   Prob(JB):                        0.813
Kurtosis:                       2.196   Cond. No.                     6.18e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.54e-34. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Test

In [8]:
X_scaled_test = scaler.transform(Xtest)

In [9]:
y_pred = results2.predict(sm.add_constant(X_scaled_test))
r2 = r2_score(Ytest, y_pred)
r2

-54.07178520418415

**4. L2, o RIDGE**

Datos TRAIN

In [10]:
ridge = Ridge(alpha=1.0)
ridge.fit(X_scaled, Ytrain)

y_pred_ridge = ridge.predict(X_scaled)

print("Coeficientes:", ridge.coef_)
print("Intercepto:", ridge.intercept_)

Coeficientes: [ 0.         -1.80358779 -1.2857849   1.01328527  1.67274528 -0.7201921
  1.26762625  0.01863527  1.40125429 -1.20813325 -0.75672773]
Intercepto: 17.433333333333334


In [11]:
r2_ridge = ridge.score(X_scaled, Ytrain)
print("R^2:", r2_ridge)

R^2: 0.9386746881405903


Datos TEST

In [12]:
ridge = Ridge(alpha=1.0)
ridge.fit(X_scaled_test, Ytest)

y_pred_ridge = ridge.predict(X_scaled_test)

print("Coeficientes:", ridge.coef_)
print("Intercepto:", ridge.intercept_)

Coeficientes: [ 0.          0.62794712  0.6879817  -1.59000081  0.47981821 -2.68884567
  0.63051843  0.36178546  0.6206204   1.53839604 -1.64658207]
Intercepto: 17.383328961740464


In [13]:
r2_ridge = ridge.score(X_scaled_test, Ytest)
print("R^2:", r2_ridge)

R^2: 0.887230499280528


**5. con 0.05 y 0.01**

Datos TRAIN, con 0.05

In [14]:
ridge2 = Ridge(alpha=0.05)
ridge2.fit(X_scaled, Ytrain)

y_pred_ridge2 = ridge2.predict(X_scaled)

print("Coeficientes:", ridge2.coef_)
print("Intercepto:", ridge2.intercept_)

Coeficientes: [ 0.         -1.93611161 -3.2764432   4.51052079  3.2845055   0.1917549
  2.45710093 -0.14580801  2.2331071  -3.50306911 -1.52384854]
Intercepto: 17.43333333333333


In [15]:
r2_ridge2 = ridge2.score(X_scaled, Ytrain)
print("R^2:", r2_ridge2)

R^2: 0.9828308657277885


Datos TEST

In [16]:
ridge2 = Ridge(alpha=0.05)
ridge2.fit(X_scaled_test, Ytest)

y_pred_ridge2 = ridge2.predict(X_scaled_test)

print("Coeficientes:", ridge2.coef_)
print("Intercepto:", ridge2.intercept_)

Coeficientes: [ 0.          1.24975801  2.53451874 -2.2600476   0.57152369 -4.2299013
  1.42432881  0.54193836  0.73388895  2.20021122 -1.2194225 ]
Intercepto: 17.41929189235009


In [17]:
r2_ridge2 = ridge2.score(X_scaled_test, Ytest)
print("R^2:", r2_ridge2)

R^2: 0.9012646685116987


Datos TRAIN, con 0.01

In [18]:
ridge3 = Ridge(alpha=0.01)
ridge3.fit(X_scaled, Ytrain)

y_pred_ridge3 = ridge3.predict(X_scaled)

print("Coeficientes:", ridge3.coef_)
print("Intercepto:", ridge3.intercept_)

Coeficientes: [ 0.         -0.65215743 -3.64173725  4.55250211  3.79292007  0.4101178
  2.49308912  0.84685452  3.38057447 -4.82957655 -1.30491028]
Intercepto: 17.43333333333333


In [19]:
r2_ridge3 = ridge3.score(X_scaled, Ytrain)
print("R^2:", r2_ridge3)

R^2: 0.9841895009460101


Datos TEST

In [20]:
ridge3 = Ridge(alpha=0.01)
ridge3.fit(X_scaled_test, Ytest)

y_pred_ridge3 = ridge3.predict(X_scaled_test)

print("Coeficientes:", ridge3.coef_)
print("Intercepto:", ridge3.intercept_)

Coeficientes: [ 0.          1.24692827  2.77644533 -2.33185289  0.56265425 -4.42245744
  1.50678084  0.55532464  0.75296246  2.21777509 -1.10734714]
Intercepto: 17.417462273257375


In [21]:
r2_ridge3 = ridge3.score(X_scaled_test, Ytest)
print("R^2:", r2_ridge3)

R^2: 0.9014244540166282


**1.2 Repite el ejercicio anterior usando 'qsec' como salida.**

In [22]:
x_qsec = motor.drop(columns=["qsec"])
y_qsec = motor["qsec"]

**1.**

In [23]:
x_qsec = sm.add_constant(x_qsec)
ols_qsec = sm.OLS(y_qsec, x_qsec)
results_qsec = ols_qsec.fit()
results_qsec.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.875
Model:                            OLS   Adj. R-squared:                  0.815
Method:                 Least Squares   F-statistic:                     14.66
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           2.44e-07
Time:                        16:48:15   Log-Likelihood:                -30.242
No. Observations:                  32   AIC:                             82.48
Df Residuals:                      21   BIC:                             98.61
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.7762      3.876      4.586      0.000       9.716      25.837
mpg            0.0690      0.061      1.123      0.274      -0.059       0.197
cyl           -0.3627      0.293     -1.239      0.229      -0.971       0.246
disp          -0.0075      0.005     -1.505      0.147      -0.018       0.003
hp            -0.0016      0.006     -0.242      0.811      -0.015       0.012
drat          -0.1311      0.476     -0.275      0.786      -1.121       0.859
wt             1.4963      0.500      2.990      0.007       0.456       2.537
vs             0.9700      0.573      1.694      0.105      -0.221       2.161
am            -0.9012      0.585     -1.540      0.139      -2.118       0.316
gear          -0.2013      0.433     -0.465      0.647      -1.101       0.699
carb          -0.2736      0.233     -1.174      0.254      -0.758       0.211
==============================================================================
Omnibus:                       21.069   Durbin-Watson:                   2.573
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               38.291
Skew:                           1.470   Prob(JB):                     4.84e-09
Kurtosis:                       7.481   Cond. No.                     8.77e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.77e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**2.**

In [24]:
Xtrain_qsec, Xtest_qsec, Ytrain_qsec, Ytest_qsec = train_test_split(x_qsec, y_qsec, train_size = 0.4, random_state = 137)

**3.**

In [25]:
scaler_qsec = StandardScaler().fit(Xtrain_qsec)
X_scaled_qsec = scaler_qsec.transform(Xtrain_qsec)

In [26]:
X_modelo2_qsec = sm.add_constant(X_scaled_qsec)
ols2_qsec = sm.OLS(Ytrain_qsec, X_modelo2_qsec)
results2_qsec = ols2_qsec.fit()
results2_qsec.summary()

/Users/juliahernandez/anaconda3/lib/python3.11/site-packages/scipy/stats/_stats_py.py:1736: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=12
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.979
Model:                            OLS   Adj. R-squared:                  0.766
Method:                 Least Squares   F-statistic:                     4.609
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.349
Time:                        16:48:15   Log-Likelihood:               -0.25847
No. Observations:                  12   AIC:                             22.52
Df Residuals:                       1   BIC:                             27.85
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.3733      0.247     70.269      0.009      14.232      20.515
x1          6.921e-15   4.79e-13      0.014      0.991   -6.09e-12     6.1e-12
x2             1.0235      2.262      0.453      0.729     -27.713      29.760
x3            -0.3081     30.092     -0.010      0.993    -382.663     382.047
x4             0.2789      1.711      0.163      0.897     -21.465      22.023
x5            -1.0127      9.720     -0.104      0.934    -124.519     122.494
x6            -1.0356      8.615     -0.120      0.924    -110.497     108.426
x7             0.5976      1.295      0.461      0.725     -15.861      17.057
x8             0.2129     23.358      0.009      0.994    -296.578     297.004
x9            -0.3877     24.867     -0.016      0.990    -316.349     315.574
x10            0.6755     25.827      0.026      0.983    -327.485     328.836
x11           -0.1562      9.422     -0.017      0.989    -119.876     119.563
==============================================================================
Omnibus:                       12.359   Durbin-Watson:                   2.150
Prob(Omnibus):                  0.002   Jarque-Bera (JB):                7.063
Skew:                           1.532   Prob(JB):                       0.0293
Kurtosis:                       5.178   Cond. No.                     4.91e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 2.73e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

In [27]:
X_scaled_test_qsec = scaler_qsec.transform(Xtest_qsec)

In [28]:
y_pred_qsec = results2_qsec.predict(sm.add_constant(X_scaled_test_qsec))
r2_qsec = r2_score(Ytest_qsec, y_pred_qsec)
r2_qsec

0.37610389620006757

**4.**

TRAIN

In [29]:
ridge_qsec = Ridge(alpha=1.0)
ridge_qsec.fit(X_scaled_qsec, Ytrain_qsec)

y_pred_ridge_qsec = ridge_qsec.predict(X_scaled_qsec)

print("Coeficientes:", ridge_qsec.coef_)
print("Intercepto:", ridge_qsec.intercept_)

Coeficientes: [ 0.          0.32243014 -0.423128   -0.09911909 -0.48174143 -0.47038265
  0.50302442  0.45486151 -0.00142143 -0.19343426 -0.27205226]
Intercepto: 17.37333333333333


In [30]:
r2_ridge_qsec = ridge_qsec.score(X_scaled_qsec, Ytrain_qsec)
print("R^2:", r2_ridge_qsec)

R^2: 0.9671802657602482


TEST

In [31]:
ridge_qsec = Ridge(alpha=1.0)
ridge_qsec.fit(X_scaled_test_qsec, Ytest_qsec)

y_pred_ridge_qsec = ridge_qsec.predict(X_scaled_test_qsec)

print("Coeficientes:", ridge_qsec.coef_)
print("Intercepto:", ridge_qsec.intercept_)

Coeficientes: [ 0.          0.23887246 -0.62032655 -0.42494672 -0.58382257 -0.00322778
  0.99668557  0.41839064 -0.64854303 -0.25409497 -0.20595853]
Intercepto: 16.97655668243855


In [32]:
r2_ridge_qsec = ridge_qsec.score(X_scaled_test_qsec, Ytest_qsec)
print("R^2:", r2_ridge_qsec)

R^2: 0.8669489775816193


TRAIN, con 0.05

In [33]:
ridge2_qsec = Ridge(alpha=0.05)
ridge2_qsec.fit(X_scaled_qsec, Ytrain_qsec)

y_pred_ridge2_qsec = ridge2_qsec.predict(X_scaled_qsec)

print("Coeficientes:", ridge2_qsec.coef_)
print("Intercepto:", ridge2_qsec.intercept_)

Coeficientes: [ 0.          0.72264822 -0.16341301  0.02886704 -0.82060213 -0.78062868
  0.68281908  0.47733881  0.00480885  0.16062277 -0.19944229]
Intercepto: 17.37333333333333


In [34]:
r2_ridge2_qsec = ridge2_qsec.score(X_scaled_qsec, Ytrain_qsec)
print("R^2:", r2_ridge2_qsec)

R^2: 0.9780489351814745


TEST

In [35]:
ridge2_qsec = Ridge(alpha=0.05)
ridge2_qsec.fit(X_scaled_test_qsec, Ytest_qsec)

y_pred_ridge2_qsec = ridge2_qsec.predict(X_scaled_test_qsec)

print("Coeficientes:", ridge2_qsec.coef_)
print("Intercepto:", ridge2_qsec.intercept_)

Coeficientes: [ 0.          0.53997486 -0.51473084 -1.07414826 -0.48157692  0.00350793
  1.67769924  0.26681069 -0.65081517 -0.31611    -0.376296  ]
Intercepto: 16.990105603448704


In [36]:
r2_ridge2_qsec = ridge2_qsec.score(X_scaled_test_qsec, Ytest_qsec)
print("R^2:", r2_ridge2_qsec)

R^2: 0.883375800821244


**2.1** Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.

1. Calcula el R2 e interpreta los signos de los betas.
2. Realiza un train-test-split donde se use el 40% de los datos para entrenar. 
3. Calcula el R2 de entrenamiento y de prueba.

**1.**

In [37]:
motor_dummies = pd.get_dummies(motor, columns = ['cyl', 'gear', 'carb'],drop_first = True)
motor_dummies.dropna()
motor_dummies.drop_duplicates()
motor_dummies.head()

mpg   disp   hp  drat     wt   qsec  vs  am  cyl_6  cyl_8  gear_4  gear_5  \
0  21.0  160.0  110  3.90  2.620  16.46   0   1      1      0       1       0   
1  21.0  160.0  110  3.90  2.875  17.02   0   1      1      0       1       0   
2  22.8  108.0   93  3.85  2.320  18.61   1   1      0      0       1       0   
3  21.4  258.0  110  3.08  3.215  19.44   1   0      1      0       0       0   
4  18.7  360.0  175  3.15  3.440  17.02   0   0      0      1       0       0   

   carb_2  carb_3  carb_4  carb_6  carb_8  
0       0       0       1       0       0  
1       0       0       1       0       0  
2       0       0       0       0       0  
3       0       0       0       0       0  
4       1       0       0       0       0

In [38]:
import pandas as pd
import numpy as np

# 1. Identificar las columnas dummy
# Esto es para que el cambio de 0 a -1 solo aplique a las columnas creadas.
columnas_a_cambiar = [col for col in motor_dummies.columns if col.startswith(('cyl', 'gear', 'carb'))]

# 2. Reemplazar los 0s por -1 en las columnas dummy
# np.where(condición, valor_si_verdadero, valor_si_falso)
for col in columnas_a_cambiar:
    motor_dummies[col] = np.where(motor_dummies[col] == 0, -1, motor_dummies[col])
    
motor_dummies.head()

mpg   disp   hp  drat     wt   qsec  vs  am  cyl_6  cyl_8  gear_4  gear_5  \
0  21.0  160.0  110  3.90  2.620  16.46   0   1      1     -1       1      -1   
1  21.0  160.0  110  3.90  2.875  17.02   0   1      1     -1       1      -1   
2  22.8  108.0   93  3.85  2.320  18.61   1   1     -1     -1       1      -1   
3  21.4  258.0  110  3.08  3.215  19.44   1   0      1     -1      -1      -1   
4  18.7  360.0  175  3.15  3.440  17.02   0   0     -1      1      -1      -1   

   carb_2  carb_3  carb_4  carb_6  carb_8  
0      -1      -1       1      -1      -1  
1      -1      -1       1      -1      -1  
2      -1      -1      -1      -1      -1  
3      -1      -1      -1      -1      -1  
4       1      -1      -1      -1      -1

In [39]:
x_dum = motor_dummies.drop(columns=["mpg"])
y_dum = motor_dummies['mpg']

In [40]:
x_dummies = pd.get_dummies(x, columns = ["cyl", "gear", "carb"])

In [41]:
X_modelo_dum = sm.add_constant(x_dum)
ols_dum = sm.OLS(y_dum, X_modelo_dum)
results_dum = ols_dum.fit()
results_dum.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.893
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     7.830
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           0.000124
Time:                        16:48:15   Log-Likelihood:                -66.608
No. Observations:                  32   AIC:                             167.2
Df Residuals:                      15   BIC:                             192.1
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         31.6279     21.399      1.478      0.160     -13.982      77.238
disp           0.0355      0.032      1.114      0.283      -0.032       0.104
hp            -0.0705      0.039     -1.788      0.094      -0.155       0.014
drat           1.1828      2.483      0.476      0.641      -4.111       6.476
wt            -4.5298      2.539     -1.784      0.095      -9.941       0.881
qsec           0.3678      0.935      0.393      0.700      -1.626       2.362
vs             1.9309      2.871      0.672      0.512      -4.189       8.051
am             1.2121      3.214      0.377      0.711      -5.637       8.062
cyl_6         -1.3243      1.520     -0.871      0.397      -4.565       1.916
cyl_8         -0.1681      3.580     -0.047      0.963      -7.798       7.462
gear_4         0.5572      1.900      0.293      0.773      -3.492       4.606
gear_5         1.2642      1.868      0.677      0.509      -2.718       5.246
carb_2        -0.4897      1.159     -0.423      0.679      -2.960       1.981
carb_3         1.4998      2.147      0.699      0.495      -3.076       6.076
carb_4         0.5457      2.225      0.245      0.810      -4.196       5.288
carb_6         2.2388      3.192      0.701      0.494      -4.565       9.042
carb_8         3.6252      4.180      0.867      0.399      -5.285      12.535
==============================================================================
Omnibus:                        2.468   Durbin-Watson:                   2.105
Prob(Omnibus):                  0.291   Jarque-Bera (JB):                1.772
Skew:                           0.576   Prob(JB):                        0.412
Kurtosis:                       3.009   Cond. No.                     1.32e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.32e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**2.**

In [42]:
Xtrain_dum, Xtest_dum, Ytrain_dum, Ytest_dum = train_test_split(x_dum, y_dum, train_size = 0.4)
#, random_state = 137

**3.**

In [43]:
scaler_dum = StandardScaler().fit(Xtrain_dum)
X_scaled_dum = scaler_dum.transform(Xtrain_dum)

TRAIN

In [44]:
ridge_train = Ridge(alpha=1.0)
ridge_train.fit(X_scaled_dum, Ytrain_dum)

y_pred = ridge_train.predict(X_scaled_dum)

print("Coeficientes:", ridge_train.coef_)
print("Intercepto:", ridge_train.intercept_)

Coeficientes: [-0.57328516  0.31302525  1.1902267  -0.72104615 -0.36893963  0.23454833
  0.24856571 -0.03180645 -0.93028114  0.0645162  -0.97376388  0.10498618
 -0.13424552 -0.90429268  0.          0.        ]
Intercepto: 17.2


In [45]:
r2_ridge_train = ridge_train.score(X_scaled_dum, Ytrain_dum)
print("R^2:", r2_ridge_train)

R^2: 0.9080341788721532


TEST

In [46]:
X_scaled_test_dum = scaler_dum.transform(Xtest_dum)

In [47]:
ridge_test = Ridge(alpha=1.0)
ridge_test.fit(X_scaled_test_dum, Ytest_dum)

y_pred_test = ridge_test.predict(X_scaled_test_dum)

print("Coeficientes:", ridge2.coef_)
print("Intercepto:", ridge2.intercept_)

Coeficientes: [ 0.          1.24975801  2.53451874 -2.2600476   0.57152369 -4.2299013
  1.42432881  0.54193836  0.73388895  2.20021122 -1.2194225 ]
Intercepto: 17.41929189235009


In [48]:
r2_ridge_test = ridge_test.score(X_scaled_dum, Ytrain_dum)
print("R^2:", r2_ridge_test)

R^2: 0.13659905160421082


**2.2** Repite el ejercicio anterior usando 'qsec' como salida.

**1.**

In [49]:
x_dum_qsec = motor_dummies.drop(columns=["qsec"])
y_dum_qsec = motor_dummies['qsec']

In [50]:
X_modelo_dum_qsec = sm.add_constant(x_dum_qsec)
ols_dum_qsec = sm.OLS(y_dum_qsec, X_modelo_dum_qsec)
results_dum_qsec = ols_dum_qsec.fit()
results_dum_qsec.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.908
Model:                            OLS   Adj. R-squared:                  0.810
Method:                 Least Squares   F-statistic:                     9.283
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           4.35e-05
Time:                        16:48:15   Log-Likelihood:                -25.252
No. Observations:                  32   AIC:                             84.50
Df Residuals:                      15   BIC:                             109.4
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.1167      5.457      2.221      0.042       0.486      23.747
mpg            0.0277      0.071      0.393      0.700      -0.123       0.178
disp           0.0035      0.009      0.389      0.703      -0.016       0.023
hp            -0.0021      0.012     -0.173      0.865      -0.027       0.023
drat           0.1079      0.687      0.157      0.877      -1.356       1.571
wt             0.8105      0.739      1.097      0.290      -0.764       2.385
vs             0.2657      0.797      0.333      0.744      -1.434       1.965
am            -1.6943      0.771     -2.197      0.044      -3.338      -0.050
cyl_6         -0.5522      0.404     -1.368      0.191      -1.412       0.308
cyl_8         -1.4835      0.905     -1.638      0.122      -3.413       0.447
gear_4         0.6661      0.494      1.348      0.198      -0.387       1.719
gear_5         0.0912      0.520      0.175      0.863      -1.018       1.200
carb_2        -0.4172      0.302     -1.384      0.187      -1.060       0.225
carb_3        -0.1147      0.598     -0.192      0.851      -1.390       1.161
carb_4        -0.9735      0.558     -1.744      0.102      -2.163       0.216
carb_6        -0.7826      0.868     -0.902      0.381      -2.632       1.067
carb_8        -0.6662      1.164     -0.572      0.576      -3.147       1.814
==============================================================================
Omnibus:                       12.619   Durbin-Watson:                   2.588
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               16.989
Skew:                           0.924   Prob(JB):                     0.000205
Kurtosis:                       6.054   Cond. No.                     1.24e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.24e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

**2.**

In [51]:
Xtrain_dum_qsec, Xtest_dum_qsec, Ytrain_dum_qsec, Ytest_dum_qsec = train_test_split(x_dum_qsec, y_dum_qsec, train_size = 0.4, random_state = 137)

**3.**

In [52]:
scaler_dum_qsec = StandardScaler().fit(Xtrain_dum_qsec)
X_scaled_dum_qsec = scaler_dum_qsec.transform(Xtrain_dum_qsec)

TRAIN

In [53]:
ridge_train_qsec = Ridge(alpha=1.0)
ridge_train_qsec.fit(X_scaled_dum_qsec, Ytrain_dum_qsec)

y_pred_qsec = ridge_train_qsec.predict(X_scaled_dum_qsec)

print("Coeficientes:", ridge_train_qsec.coef_)
print("Intercepto:", ridge_train_qsec.intercept_)

Coeficientes: [ 0.23537669  0.15717246 -0.64365739 -0.31726369  0.42737004  0.5428533
  0.01976446 -0.15416478 -0.30527997  0.0495953  -0.29568849 -0.15045977
 -0.06771537 -0.42914262  0.         -0.18708807]
Intercepto: 17.37333333333333


In [54]:
r2_ridge_train_qsec = ridge_train_qsec.score(X_scaled_dum_qsec, Ytrain_dum_qsec)
print("R^2:", r2_ridge_train_qsec)

R^2: 0.9785215603421599


TEST

In [55]:
X_scaled_test_dum_qsec = scaler_dum_qsec.transform(Xtest_dum_qsec)

In [56]:
ridge_test_qsec = Ridge(alpha=1.0)
ridge_test_qsec.fit(X_scaled_test_dum_qsec, Ytest_dum_qsec)

y_pred_qsec_test = ridge_test_qsec.predict(X_scaled_test_dum_qsec)

print("Coeficientes:", ridge_test_qsec.coef_)
print("Intercepto:", ridge_test_qsec.intercept_)

Coeficientes: [ 1.99244780e-01 -1.51209005e-01 -3.58621578e-01  2.57919282e-02
  7.64242138e-01  3.80786639e-01 -7.61698092e-01 -3.43146411e-01
 -8.26042680e-01  3.09078124e-01 -1.73024751e-01 -5.08263440e-02
  5.67876211e-02 -4.74278055e-01 -1.25930424e-01  3.20516357e-31]
Intercepto: 17.06381259299005


In [57]:
r2_ridge_test_qsec = ridge_test_qsec.score(X_scaled_test_dum_qsec, Ytest_dum_qsec)
print("R^2:", r2_ridge_test_qsec)

R^2: 0.8912788065959261


**3.1 Compara los R2 de los ejercicios 1.1 & 2.1.**

Los R^2 fueron los siguientes:

En el ejercicio 1.1: 

- 0.869 con todos los datos
- 0.990 con datos de TRAIN
- -54 con datos de TEST

En el ejercicio 2.1:

- 0.893 con todos los datos
- 0.917 con datos de TRAIN 
- 0.583 con datos de TEST



El modelo es mucho más confiable cuando se utiliza 'qsec' como datos de Y ya que, comparando los R^2 de los datos de test tanto con 'qsec' como con 'mpg', el resultado es mucho mejor con 'qsec' a pesar de que en sí no fue tan bueno

**3.2 Compara los R2 de los ejercicios 1.2 & 2.2.**

Los R^2 fueron los siguientes:

En el ejercicio 1.2: 

- 0.875 con todos los datos
- 0.979 con datos de TRAIN
- 0.376 con datos de TEST

En el ejercicio 2.2:

- 0.908 con todos los datos
- 0.978 con datos de TRAIN 
- 0.891 con datos de TEST

Comparando los 4 resultados del R^2, puedo concluir que utilizar dummies es más confiable que no hacerlo, porque tanto con 'qsec' como con 'mpg', los resultados a la hora de utilizar los datos de TEST fueron más altos que cuando no utilicé dummies. Además, al igual que con la conclusión pasada, utilizar 'qsec' como datos de Y es más confiable